### PR TITLE
[SPARK-15355] [CORE] Proactive block replication

### PR DESCRIPTION
**Pull request by shubhamchopra**
## What changes were proposed in this pull request?

We are proposing addition of pro-active block replication in case of executor failures. BlockManagerMasterEndpoint does all the book-keeping to keep a track of all the executors and the blocks they hold. It also keeps a track of which executors are alive through heartbeats. When an executor is removed, all this book-keeping state is updated to reflect the lost executor. This step can be used to identify executors that are still in possession of a copy of the cached data and a message could be sent to them to use the existing "replicate" function to find and place new replicas on other suitable hosts. Blocks replicated this way will let the master know of their existence.

This can happen when an executor is lost, and would that way be pro-active as opposed be being done at query time.
## How was this patch tested?

This patch was tested with existing unit tests along with new unit tests added to test the functionality.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/storage/BlockManagerId.scala [hunk 1]
```diff
@@ -37,10 +37,11 @@ import org.apache.spark.util.Utils
 class BlockManagerId private (
     private var executorId_ : String,
     private var host_ : String,
-    private var port_ : Int)
+    private var port_ : Int,
+    private var topologyInfo_ : Option[String])
```
Comment (rxin) [inline core/src/main/scala/org/apache/spark/storage/BlockManagerId.scala]: we need to document what information this parameter contains (and in what format).

Comment (shubhamchopra) [inline core/src/main/scala/org/apache/spark/storage/BlockManagerId.scala]: Added documentation about the parameter in the constructor.

**Comment by sameeragarwal** [comment]
jenkins ok to test

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala [hunk 2]
```diff
@@ -1152,20 +1185,25 @@ private[spark] class BlockManager(
...
+      initialPeers,
+      peersReplicatedTo,
       blockId,
       numPeersToReplicateTo)
 
     while(numFailures <= maxReplicationFailures &&
-        !peersForReplication.isEmpty &&
-        peersReplicatedTo.size != numPeersToReplicateTo) {
+      !peersForReplication.isEmpty &&
+      peersReplicatedTo.size < numPeersToReplicateTo) {
```
Comment (sameeragarwal) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: While I think it's still valid to replace the inequality with a strictly-less-than check, but just out of curiosity, can the number of `peersReplicatedTo` ever exceed `numPeersToReplicateTo`?

File: core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala [hunk 3]
```diff
@@ -188,24 +189,45 @@ class BlockManagerMasterEndpoint(
   }
 
   private def removeBlockManager(blockManagerId: BlockManagerId) {
+    val proactivelyReplicate = conf.get("spark.storage.replication.proactive", "false").toBoolean
```
Comment (sameeragarwal) [inline core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala]: can probably move this out of the function block

File: core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala [hunk 4]
```diff
@@ -188,24 +189,45 @@ class BlockManagerMasterEndpoint(
...
+
     val info = blockManagerInfo(blockManagerId)
 
     // Remove the block manager from blockManagerIdByExecutor.
     blockManagerIdByExecutor -= blockManagerId.executorId
 
-    // Remove it from blockManagerInfo and remove all the blocks.
-    blockManagerInfo.remove(blockManagerId)
```
Comment (sameeragarwal) [inline core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala]: why did you move this at the end?

File: core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala [hunk 5]
```diff
@@ -188,24 +189,45 @@ class BlockManagerMasterEndpoint(
...
     val iterator = info.blocks.keySet.iterator
     while (iterator.hasNext) {
       val blockId = iterator.next
       val locations = blockLocations.get(blockId)
       locations -= blockManagerId
       if (locations.size == 0) {
         blockLocations.remove(blockId)
+        logWarning(s"No more replicas available for $blockId !")
+      } else if ((blockId.isRDD || blockId.isInstanceOf[TestBlockId]) && proactivelyReplicate) {
+        // we only need to proactively replicate RDD blocks
```
Comment (sameeragarwal) [inline core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala]: I think it makes sense overall but it'd be great to add some more comments about why are we only concerned with RDD blocks and not others.

Comment (sameeragarwal) [inline core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala]: Also, might be nicer to make `proactivelyReplicate` the first check for better short-circuiting.

File: core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala [hunk 6]
```diff
@@ -188,24 +189,45 @@ class BlockManagerMasterEndpoint(
...
+        // we also need to replicate this behavior for test blocks for unit tests
+        // we send a message to a randomly chosen executor location to replicate block
+        // assuming single executor failure, we find out how many replicas existed before failure
+        val maxReplicas = locations.size + 1
```
Comment (sameeragarwal) [inline core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala]: What happens if multiple executors are removed simultaneously? Depending on the invocation sequence, is it possible for `maxReplicas` to be significantly less than the original number of replicas?

**Review by sameeragarwal — COMMENTED** [review]
Thanks @shubhamchopra, I left a few comments and questions but overall this approach looks quite reasonable to me.

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala [hunk 7]
```diff
@@ -1131,14 +1131,47 @@ private[spark] class BlockManager(
...
 
   /**
+   * Called for pro-active replenishment of blocks lost due to executor failures
+   *
+   * @param blockId blockId being replicate
+   * @param replicas existing block managers that have a replica
+   * @param maxReps maximum replicas needed
+   * @return
+   */
+  def replicateBlock(blockId: BlockId, replicas: Set[BlockManagerId], maxReps: Int): Boolean = {
```
Comment (shubhamchopra) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: This doesn't need to return a boolean. Changing the return type to Unit. Also changing the variable names.

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala [hunk 8]
```diff
@@ -1131,14 +1131,47 @@ private[spark] class BlockManager(
...
+
+  /**
    * Replicate block to another node. Note that this is a blocking call that returns after
    * the block has been replicated.
+   *
+   * @param blockId
+   * @param data
+   * @param level
+   * @param classTag
+   * @param existingReplicas
```
Comment (shubhamchopra) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: Removing these.

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala [hunk 9]
```diff
@@ -1152,20 +1185,25 @@ private[spark] class BlockManager(
 
     val startTime = System.nanoTime
 
-    var peersReplicatedTo = mutable.HashSet.empty[BlockManagerId]
+    var peersReplicatedTo = mutable.HashSet.empty ++ existingReplicas
     var peersFailedToReplicateTo = mutable.HashSet.empty[BlockManagerId]
     var numFailures = 0
 
+    val initialPeers = {
```
Comment (shubhamchopra) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: Done

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala (hunk 2 above)
Comment (shubhamchopra) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: One scenario I can think of is if an executor with the block being replicated is lost (due to say a delayed heartbeat) and joins back again. The current implementation would recognize the block manager needs to reregister and will report all blocks. The probability of this happening increases with pro-active replication, I think.

File: core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala (hunk 3 above)
Comment (shubhamchopra) [inline core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala]: Removed from the function.

File: core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala (hunk 6 above)
Comment (shubhamchopra) [inline core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala]: Yes, that's a tough one. So the way replication is implemented, the correct storage level is only available with one of the blocks at BlockManager layer (we don't have access to RDD that this block is a part of, so we can't extract information from there). The remaining blocks all have storage levels set to 1. So I use the locations size to get an approximation for the storage level.

File: core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala [hunk 10]
```diff
@@ -195,17 +198,39 @@ class BlockManagerMasterEndpoint(
...
     val iterator = info.blocks.keySet.iterator
     while (iterator.hasNext) {
       val blockId = iterator.next
       val locations = blockLocations.get(blockId)
       locations -= blockManagerId
       if (locations.size == 0) {
         blockLocations.remove(blockId)
+        logWarning(s"No more replicas available for $blockId !")
+      } else if (proactivelyReplicate && (blockId.isRDD || blockId.isInstanceOf[TestBlockId])) {
+        // only RDD blocks store data that users explicitly cache so we only need to proactively
```
Comment (sameeragarwal) [inline core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala]: Just a nit, but how about we split this comment into two separate ones for better readability around the two separate set of issues. My suggestion:

```scala
      // De-register the block if none of the block managers have it. Otherwise, if pro-active
      // replication is enabled, and a block is either an RDD or a test block (the latter is used
      // for unit testing), we send a message to a randomly chosen executor location to replicate
      // the given block. Note that we ignore other block types (such as broadcast/shuffle blocks
      // etc.) as replication doesn't make much sense in that context.
      if (locations.size == 0) {
        blockLocations.remove(blockId)
      } else if (proactivelyReplicate && (blockId.isRDD || blockId.isInstanceOf[TestBlockId])) {
      // As a heursitic, assume single executor failure to find out the number of replicas that
      // existed before failure
      val maxReplicas = locations.size + 1

      val i = (new Random(blockId.hashCode)).nextInt(locations.size)
      val blockLocations = locations.toSeq
      val candidateBMId = blockLocations(i)
      val blockManager = blockManagerInfo.get(candidateBMId)
      if(blockManager.isDefined) {
        val remainingLocations = locations.toSeq.filter(bm => bm != candidateBMId)
        val replicateMsg = ReplicateBlock(blockId, remainingLocations, maxReplicas)
        blockManager.get.slaveEndpoint.ask[Boolean](replicateMsg)
      }
    }
```

**Review by sameeragarwal — APPROVED** [review]
LGTM, modulo some minor nits. @joshRosen can you please take a look too?

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala [hunk 11]
```diff
@@ -1131,14 +1131,43 @@ private[spark] class BlockManager(
...
   /**
+   * Called for pro-active replenishment of blocks lost due to executor failures
+   *
+   * @param blockId blockId being replicate
+   * @param existingReplicas existing block managers that have a replica
+   * @param maxReplicas maximum replicas needed
+   * @return
+   */
+  def replicateBlock(
+    blockId: BlockId,
```
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: Same as Sameer's comment elsewhere in the code, we should fix the indentation here.

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala [hunk 12]
```diff
@@ -1131,14 +1131,43 @@ private[spark] class BlockManager(
   }
 
   /**
+   * Called for pro-active replenishment of blocks lost due to executor failures
+   *
+   * @param blockId blockId being replicate
+   * @param existingReplicas existing block managers that have a replica
+   * @param maxReplicas maximum replicas needed
+   * @return
```
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: You can omit this `@return` since this method doesn't have a return value.

File: core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala [hunk 13]
```diff
@@ -195,17 +198,39 @@ class BlockManagerMasterEndpoint(
...
+        // we also need to replicate this behavior for test blocks for unit tests
+        // we send a message to a randomly chosen executor location to replicate block
+        // assuming single executor failure, we find out how many replicas existed before failure
+        val maxReplicas = locations.size + 1
+
+        val i = (new Random(blockId.hashCode)).nextInt(locations.size)
+        val blockLocations = locations.toSeq
+        val candidateBMId = blockLocations(i)
+        val blockManager = blockManagerInfo.get(candidateBMId)
+        if(blockManager.isDefined) {
```
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala]: Nit: space after `if`.

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala [hunk 14]
```diff
@@ -1131,14 +1131,43 @@ private[spark] class BlockManager(
...
+   */
+  def replicateBlock(
+    blockId: BlockId,
+    existingReplicas: Set[BlockManagerId],
+    maxReplicas: Int): Unit = {
+    logInfo(s"Pro-actively replicating $blockId")
+    val infoForReplication = blockInfoManager.lockForReading(blockId).map { info =>
+      val data = doGetLocalBytes(blockId, info)
+      val storageLevel = StorageLevel(
+        info.level.useDisk,
```
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: Minor nit, but a problem with the `StorageLevel` constructor is that it has a bunch of adjacent boolean parameters, so in such cases I'd usually prefer to name all of the parameters explicitly at the call site in order to avoid errors should these lines get permuted / to convince readers that the API is being used correctly.

Thus I'd probably write each line like `useDisk = info.level.useDisk,` etc.

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala [hunk 15]
```diff
@@ -1131,14 +1131,43 @@ private[spark] class BlockManager(
...
+   * @param existingReplicas existing block managers that have a replica
+   * @param maxReplicas maximum replicas needed
+   * @return
+   */
+  def replicateBlock(
+    blockId: BlockId,
+    existingReplicas: Set[BlockManagerId],
+    maxReplicas: Int): Unit = {
+    logInfo(s"Pro-actively replicating $blockId")
+    val infoForReplication = blockInfoManager.lockForReading(blockId).map { info =>
```
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: This call acquires a read lock on the block, but when is that lock released? Per the Scaladoc of `doGetLocalBytes`, you need to be holding a read lock before calling that method, but upon successful return from that method the read lock will still be held by the caller.

I think what you want to do is acquire the lock, immediately call `doGetLocalBytes`, then begin a `try-finally` statement to call `replicate()` and unlock / release the lock in the `finally` block.

Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: Also, I don't think there's a need to have separate `.map` and `.foreach` calls over the option. Instead, I think it would be clearer to avoid the assignment to the `infoForReplication` variable and just perform all of the work inside of a `.foreach` call on the `Option` with the block info.

File: core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala (hunk 10 above)
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala]: +1 on Sameer's suggestions. This code is a little subtle and benefits from a clearer comment.

File: core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala [hunk 16]
```diff
@@ -195,17 +198,39 @@ class BlockManagerMasterEndpoint(
...
+      } else if (proactivelyReplicate && (blockId.isRDD || blockId.isInstanceOf[TestBlockId])) {
+        // only RDD blocks store data that users explicitly cache so we only need to proactively
+        // replicate RDD blocks
+        // broadcast related blocks exist on all executors, so we don't worry about them
+        // we also need to replicate this behavior for test blocks for unit tests
+        // we send a message to a randomly chosen executor location to replicate block
+        // assuming single executor failure, we find out how many replicas existed before failure
+        val maxReplicas = locations.size + 1
+
+        val i = (new Random(blockId.hashCode)).nextInt(locations.size)
```
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala]: Why do we need to use a fixed random seed here? Testing?

Also, isn't there a `Random.choice()` that you can use for this? Or a method like that in our own `Utils` class?

File: core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala (hunk 13 above)
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala]: If you're not going to have an `else` branch here then you might as well just `forEach` over the result of `blockManagerInfo.get(candidateBMId)`.

File: core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala [hunk 17]
```diff
@@ -195,17 +198,39 @@ class BlockManagerMasterEndpoint(
...
+        val blockLocations = locations.toSeq
+        val candidateBMId = blockLocations(i)
+        val blockManager = blockManagerInfo.get(candidateBMId)
+        if(blockManager.isDefined) {
+          val remainingLocations = locations.toSeq.filter(bm => bm != candidateBMId)
```
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala]: Is it possible for this list to be empty in certain corner-cases? What happens if `ReplicateBlock` is called with an empty set of locations? Is it just a no-op in that case?

File: core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala [hunk 18]
```diff
@@ -65,6 +66,8 @@ class BlockManagerMasterEndpoint(
     mapper
   }
 
+  val proactivelyReplicate = conf.get("spark.storage.replication.proactive", "false").toBoolean
```
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala]: Please document this new configuration in `docs/configuration.md`.

**Review by JoshRosen — CHANGES_REQUESTED** [review]
I just took a review pass. I have some concerns about the release of the readLock which is acquired in the replication code. Unless I'm overlooking something, I think it's not currently released following the proactive replication.

I also left several small comments on how I think the code could be slightly simplified. I'll also echo Sameer's review feedback regarding minor code style issues and clarification of some of the comments.

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala [hunk 19]
```diff
@@ -1131,14 +1131,47 @@ private[spark] class BlockManager(
...
+   * @param data
+   * @param level
+   * @param classTag
+   * @param existingReplicas
    */
   private def replicate(
-      blockId: BlockId,
-      data: ChunkedByteBuffer,
-      level: StorageLevel,
-      classTag: ClassTag[_]): Unit = {
```
Comment (shubhamchopra) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: @sameeragarwal This code is being removed as a part of this PR. Code replacing this has this fixed.

File: core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala (hunk 10 above)
Comment (shubhamchopra) [inline core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala]: Agree with the suggestion. Modifying it. Thanks!

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala (hunk 15 above)
Comment (shubhamchopra) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: Yes, that makes sense.

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala (hunk 14 above)
Comment (shubhamchopra) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: Done

File: core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala (hunk 16 above)
Comment (shubhamchopra) [inline core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala]: Scala Random api doesn't have a choice method. And Spark Utils class has methods to shuffle, but not a random choice.

File: core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala (hunk 17 above)
Comment (shubhamchopra) [inline core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala]: If we are at this point, there would be atleast one location with the block which will get chosen as the candidate here. `remainingLocations` just tells the replication logic where other replicas are present (if any, so it can be an empty set), so it can use that info while choosing candidate executors for replication.

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala (hunk 15 above)
Comment (sameeragarwal) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: Nice catch! Can we also assert that all locks are released somewhere in `testProactiveReplication`?

**Review by JoshRosen — APPROVED** [review]
Thanks for incorporating my feedback. LGTM.

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala (hunk 15 above)
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: I think that we can set `spark.storage.exceptionOnPinLeak` to `true` in SparkConf to do this.

**Comment by shubhamchopra** [comment]
"spark.storage.exceptionOnPinLeak" based check only works if executors are created. I put in an assertion check using logic similar to it in the testProactiveReplication tests.

**Review by sameeragarwal — APPROVED** [review]
Looks good, thanks!

**Comment by cloud-fan** [comment]
thanks, merging to master!
